### PR TITLE
Fix member list rendering after removal

### LINKED ISSUE #18851: [BUG] Removing a member can display the wrong remaining member
**Setup**
- Rancher version: v2.16.0 (master)
- Rancher UI Extensions: None
- Browser type & version: Google Chrome (latest)

**Describe the bug**

When removing a member from the middle of the Project Members or Cluster Membership list, the UI can display the wrong remaining members.

For example, when members 1, 2, 3, 4, and 5 are listed and member 4 is removed, member 5 may appear to have been removed instead. Saving still removes the correct member, so the persisted data is correct.

**To Reproduce**

1. Open a Project Members or Cluster Membership editor.
2. Add at least five members.
3. Remove a member that is not the last member.
4. Observe the displayed member list before saving.

The issue can also occur when the same principal is added multiple times.

**Result**

A different member can appear to have been removed from the list.

**Expected Result**

The selected member should disappear immediately, and all other members should remain visible in the correct order.

**Additional context**

This is a UI rendering issue. The correct membership change is persisted after saving.

Related PR: #18642

### PR DESCRIPTION
**Pull request by hwanyoungChoi**
### Summary

Fixes #18851

Fix an issue where removing a member from the middle of the member list could make the next member appear to disappear in the UI before saving.

### Occurred changes and/or fixed issues

- Added support for an optional stable row key in `ArrayList`.
- Passed a stable key for membership bindings in `MembershipEditor`.
- Removing the fourth member from a five-member list now immediately displays members 1, 2, 3, and 5.
- The fix applies to editors that use `MembershipEditor`, including Project Members and Cluster Membership.

### Technical notes summary

`ArrayList` used the row index as the Vue key. After a member was removed, Vue reused the following row's component instance. This could leave `Principal` displaying cached data for the removed member position.

`MembershipEditor` now provides a stable key based on the binding ID when available, or the principal and role template identifiers for unsaved bindings.

Added a regression test covering removal of a member in the middle of the list.

### Areas or cases that should be tested

1. Open a Cluster or Project membership editor with at least five members.
2. Remove a member that is not the last member.
3. Verify that the selected member disappears immediately and all remaining members stay visible in the correct order.
4. Save the changes.
5. Reload the page and verify that the saved membership list matches the displayed list.

Automated tests run:

```text
yarn test:ci shell/components/form/Members/__tests__/MembershipEditor.test.ts shell/components/form/Members/__tests__/MembershipEditor.remove.test.ts shell/components/form/__tests__/ArrayList.test.ts
```

### Areas which could experience regressions

- Project and Cluster membership editing flows.
- Adding and removing unsaved membership bindings.
- Other ArrayList consumers that opt in to rowKey.

### Screenshot/Video
Not applicable. This change corrects row rendering after a removal action.

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone 
- [x] The PR template has been filled out
- [x] The PR has been self reviewed 
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes
- [x] The PR has been reviewed in terms of Accessibility
- [x] The PR has considered, and if applicable tested with, the three Global Roles `Admin`, `Standard User` and `User Base`

### CODE REVIEW

File: shell/components/form/ArrayList.vue [hunk 1]
```diff
@@ -361,7 +365,7 @@ export default {
         </div>
         <div
           v-for="(row, idx) in rows"
-          :key="idx"
+          :key="rowKey ? rowKey(row.value, idx) : idx"
```
Comment (rak-phillip) [inline shell/components/form/ArrayList.vue]: We can end up in a scenario with duplicate keys. For example, if we create users `test01`-`test05` and then add `test01` multiple times as a cluster member, we will generate multiple keys with the same value: `local://m-afdk7-cluster-member`

Image

We will want a better guarantee that the key is unique if we are going to move away from the invalid pattern of using the index[^1].

> Children of the same common parent must have unique keys. Duplicate keys will cause render errors.

[^1]: https://vuejs.org/api/built-in-special-attributes#key

File: shell/components/form/ArrayList.vue [hunk 2]
```diff
@@ -115,6 +115,10 @@ export default {
       type:    String,
       default: 'array-list',
     },
+    rowKey: {
+      type:    Function,
```
Comment (rak-phillip) [inline shell/components/form/ArrayList.vue]: I have some concerns with using `Function` as a type for the `rowKey` prop. The function contract isn't documented with the current implementation. Nothing states the arity, the return type, or the uniqueness requirements of `rowKey`. Can we minimally add a documentation block if the function is absolutely required for this prop? Are there alternatives that we can consider for producing a unique key when one is required?

**Review by rak-phillip — CHANGES_REQUESTED** [review]
@hwanyoungChoi Thanks for raising this PR and helping to make Rancher better! 

I was able to give this a review and I think that we will want follow-up with some updates before we move ahead with this change.

I'm also noting the "The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed" option is checked, but I don't see an associated issue linked to this change. Is there a specific issue that you are targeting with this fix? I think that we will need one for additional testing after we merge a fix.

File: shell/components/form/ArrayList.vue (hunk 2 above)
Comment (hwanyoungChoi) [reply shell/components/form/ArrayList.vue]: In this revision, the `rowKey` prop was removed.

Instead of adding a new rule to the shared component, the necessary handling was implemented only within the specific member list screen where the issue occurred.

File: shell/components/form/ArrayList.vue (hunk 1 above)
Comment (hwanyoungChoi) [reply shell/components/form/ArrayList.vue]: The previous method generated keys by combining the user ID and role, which could lead to key collisions if the same user was added multiple times.

Now, keys are generated based on the specific member registration details, ensuring that unique keys are used even when the same user is added repeatedly.

I have also added a test case to verify that the list displays correctly when one instance of a duplicate user is removed.

**Comment by hwanyoungChoi** [comment]
I selected the checklist item by mistake and have unchecked it. There is no existing issue specifically addressing this member list rendering problem.

Should I create a new issue for QA tracking and link it to this PR?

**Comment by rak-phillip** [comment]
> Should I create a new issue for QA tracking and link it to this PR?

If you could, please create one. It will help us to properly track the issue for the upcoming 2.16 release.

**Comment by hwanyoungChoi** [comment]
> > Should I create a new issue for QA tracking and link it to this PR?
> 
> If you could, please create one. It will help us to properly track the issue for the upcoming 2.16 release.

I created #18851 and linked it to this PR for tracking.

**Review by rak-phillip — CHANGES_REQUESTED** [review]
This looks good. There are just a few type-errors that need to be resolved before we can merge:

```
  ✖ 3 new type error(s) introduced:
  
    shell/components/form/Members/__tests__/MembershipEditor.remove.test.ts: error TS2322: Type '() => { schema: null; lastSavedBindings: never[]; bindings: { principalId: string; roleDisplay: string; roleTemplateId: string; }[]; }' is not assignable to type '() => Partial<{ schema: any; bindings: never[]; lastSavedBindings: never[]; }>'.
    shell/components/form/Members/__tests__/MembershipEditor.remove.test.ts: error TS2339: Property 'value' does not exist on type '{ props: string[]; data(): any; template: string; }'.
    shell/components/form/Members/__tests__/MembershipEditor.remove.test.ts: error TS7023: 'data' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.
```

https://github.com/rancher/dashboard/actions/runs/31466055576/job/94165830478?pr=18642

**Review by rak-phillip — APPROVED** [review]
LGTM
